File: day04/security.pl
#!/usr/bin/env perl
use strict;
use warnings;
use 5.24.0;
use feature 'signatures';
no warnings "experimental::signatures";

my $sum = 0;

while (<>) {
    chomp;
    m/([a-z\-]+)-(\d+)\[([^]]+)/;
    my ($name, $sector, $checksum) = ($1, $2, $3);
    if (real_room($name, $checksum)) {
        $sum += $sector;
        my $decoded = decode($name, $sector);
        say "$decoded => $_";
    }
}

say $sum;

sub real_room($name, $checksum) {
    my %cnt;

    $name =~ s/-//g;
    $cnt{$_}++ for split //, $name;

    my @s = sort {$cnt{$b} <=> $cnt{$a} || $a cmp $b} keys %cnt;
    if (@s < 5) {
        return 0;
    } else {
        return $checksum eq join "", @s[0..4];
    }
}

sub decode($name, $sector) {
    my @decoded = map { rotate($_, $sector) } split //, $name;
    return join "", @decoded;
}

sub rotate($s, $cnt) {
    if ($s eq '-') {
        return ' ';
    } else {
        my $n = ord($s) - ord('a') + $cnt;
        $n %= 26;
        return chr($n + ord('a'));
    }
}

File: day04/security.py
#!/usr/bin/env python
from sys import argv
import re
import collections

def real_room(name, checksum) :
    cnt = collections.defaultdict(int)

    for s in name :
        if s != '-' :
            cnt[s] = cnt[s] + 1

    sorted_keys = sorted(cnt.keys(), cmp=lambda x,y: cmp(cnt[y],cnt[x]) or cmp(x,y))
    if len(sorted_keys) < 5 :
        return False
    else :
        return checksum == "".join(sorted_keys[0:5])

def decode(name, sector) :
    decoded = ""
    for c in name :
        decoded = decoded + rotate(c, sector)

    return decoded

def rotate(c, cnt) :
    if c == "-" :
        return " "
    else :
        n = ord(c) - ord("a") + int(cnt)
        n = n % 26
        return chr(n + ord("a"))

total = 0

script, filename = argv
with open(filename) as f :
    for line in f :
        line = line.rstrip("\n")
        m = re.match('([a-z\-]+)-(\d+)\[([^]]+)', line)
        (name, sector, checksum) = m.group(1, 2, 3)
        if real_room(name, checksum) :
            total = total + int(sector)
            decoded = decode(name, sector)
            print "%s => %s" % (decoded, line)

print total


